pkg/infoschema/infoschema_v2.go

```diff
@@ -0,0 +1,431 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package infoschema
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/pkg/kv"
+	"github.com/pingcap/tidb/pkg/meta"
+	"github.com/pingcap/tidb/pkg/meta/autoid"
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/table"
+	"github.com/scalalang2/golang-fifo"
+	"github.com/scalalang2/golang-fifo/sieve"
+	"github.com/tidwall/btree"
+	"golang.org/x/sync/singleflight"
+)
+
+var enableV2 atomic.Bool
+
+// Item is the btree item sorted by name or by id.
+type Item struct {
+	dbName        string
+	dbID          int64
+	tableName     string
+	tableID       int64
+	schemaVersion int64
+}
+
+type schemaItem struct {
+	schemaVersion int64
+	dbInfo        *model.DBInfo
+}
+
+func (si *schemaItem) Name() string {
+	return si.dbInfo.Name.L
+}
+
+// versionAndTimestamp is the tuple of schema version and timestamp.
+type versionAndTimestamp struct {
+	schemaVersion int64
+	timestamp     uint64
+}
+
+// Data is the core data struct of infoschema V2.
+type Data struct {
+	// For the TableByName API, sorted by {dbName, tableName, tableID} => schemaVersion
+	//
+	// If the schema version +1 but a specific table does not change, the old record is
+	// kept and no new {dbName, tableName, tableID} => schemaVersion+1 record been added.
+	//
+	// It means as long as we can find a item in it, the item is available, even through the
+	// schema version maybe smaller than required.
+	//
+	// *IMPORTANT RESTRICTION*: Do we have the full data in memory? NO!
+	name2id *btree.BTreeG[Item]
+
+	// For the TableByID API, sorted by {tableID}
+	// To reload model.TableInfo, we need both table ID and database ID for meta kv API.
+	// It provides the tableID => databaseID mapping.
+	//
+	// *IMPORTANT RESTRICTION*: Do we have the full data in memory? NO!
+	// But this mapping should be synced with name2id.
+	byID *btree.BTreeG[Item]
+
+	cache fifo.Cache[cacheKey, table.Table]
+
+	// For the SchemaByName API
+	schemaMap *btree.BTreeG[schemaItem]
```

Comment:
maybe add comments about how it's sorted as previous 2 fields do

Reply:
Done